File: server-ruby-sinatra/api/game_level_api.rb
require 'json'


MyApp.add_route('POST', '/api/{version}/level/create', {
  "resourcePath" => "/GameLevel",
  "summary" => "Create Game Level",
  "nickname" => "create_game_level",
  "responseClass" => "GameLevelResponse",
  "endpoint" => "/api/{version}/level/create",
  "notes" => "Create a game level. Currently does NOT support game objects.",
  "parameters" => [
    {
      "name" => "account_id",
      "description" => "The logged in user.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "app_key",
      "description" => "The game application key to save the level for.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "name",
      "description" => "The name of the level.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "description",
      "description" => "The description of the level.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "difficulty",
      "description" => "The difficulty, possible values are: VERY_EASY, EASY, MEDIUM, HARD, VERY_HARD.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "app_version",
      "description" => "The version number of the application required to correctly load/play the level.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "asset_image_id",
      "description" => "The asset Id of the level image.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "asset_icon_id",
      "description" => "The asset Id of the level icon.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "game_data",
      "description" => "The game level data: xml, json, or other text based format.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "game_data_suffix",
      "description" => "The game level data format type.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "visibility",
      "description" => "Is the level visible to others, possible values are: PUBLIC, PRIVATE.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "friend_group",
      "description" => "Make the level be readable by all friends.",
      "dataType" => "Boolean",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "connection_ids",
      "description" => "Make the level be readable by connections in this list.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "connection_group_ids",
      "description" => "Make the level be readable by connection groups in this list.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "balance",
      "description" => "Set the amount of money available to spend, once 0 deactivate level. Set to a negative value for unlimited.",
      "dataType" => "Float",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "active",
      "description" => "If true set the game level as active. Default is false.",
      "dataType" => "Boolean",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "allocate_tickets",
      "description" => "If true then scoring will give tickets. Default is false.",
      "dataType" => "Boolean",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "ticket_count",
      "description" => "The number of tickets to reward",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "ticket_type",
      "description" => "The type of ticket to reward, null means default type",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "points",
      "description" => "The number of points to award for completing a level",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "tutorial_title",
      "description" => "Title of the tutorial.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "tutorial_message",
      "description" => "Message of the tutotrial.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "tutorial_alignment",
      "description" => "Alignment of the tutorial image. Default to NONE. Possible values are: NONE, IMAGE_ABOVE, IMAGE_BELOW, IMAGE_LEFT, IMAGE_RIGHT, IMAGE_ONLY, TEXT_ONLY",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "tutorial_image_asset_id",
      "description" => "Asset id of the tutorial image.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "offer_id",
      "description" => "id of the offer",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "meta_data",
      "description" => "external custom client defined data",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "version",
      "description" => "",
      "dataType" => "Float",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('POST', '/api/{version}/level/delete', {
  "resourcePath" => "/GameLevel",
  "summary" => "Delete Game Level",
  "nickname" => "delete_game_level",
  "responseClass" => "SirqulResponse",
  "endpoint" => "/api/{version}/level/delete",
  "notes" => "Delete a game level. The level and account must be valid and have the appropirate permissions to view the content.",
  "parameters" => [
    {
      "name" => "account_id",
      "description" => "The logged in user.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "level_id",
      "description" => "The id of the level to return.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "version",
      "description" => "",
      "dataType" => "Float",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('GET', '/api/{version}/level/get', {
  "resourcePath" => "/GameLevel",
  "summary" => "Get Game Level",
  "nickname" => "get_game_level",
  "responseClass" => "GameLevelResponse",
  "endpoint" => "/api/{version}/level/get",
  "notes" => "Get a game level. The level and account must be valid and have the appropirate permissions to view the content.",
  "parameters" => [
    {
      "name" => "account_id",
      "description" => "The logged in user.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "level_id",
      "description" => "The id of the level to return.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "include_game_data",
      "description" => "If true include the game level data, otherwise don&#39;t. default is false.",
      "dataType" => "Boolean",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "version",
      "description" => "",
      "dataType" => "Float",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('GET', '/api/{version}/level/search', {
  "resourcePath" => "/GameLevel",
  "summary" => "Search Game Levels",
  "nickname" => "get_game_levels_by_application",
  "responseClass" => "GameLevelListResponse",
  "endpoint" => "/api/{version}/level/search",
  "notes" => "Get a list of levels for an application, just those the account has permissions to view.",
  "parameters" => [
    {
      "name" => "account_id",
      "description" => "The logged in user.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "app_key",
      "description" => "the application key",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "keyword",
      "description" => "Match the keyword to the owner name or level name.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "sort_field",
      "description" => "what field to sort on",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "descending",
      "description" => "whether to return levels in ascending or descending order",
      "dataType" => "Boolean",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "start",
      "description" => "Start the result set at some index.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "limit",
      "description" => "Limit the result to some number.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "app_version",
      "description" => "The maximum version of the level to return.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "include_game_data",
      "description" => "If true include the game level data, otherwise don&#39;t. default is false.",
      "dataType" => "Boolean",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "filters",
      "description" => "",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "version",
      "description" => "",
      "dataType" => "Float",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('GET', '/api/{version}/level/searchByBillableEntity', {
  "resourcePath" => "/GameLevel",
  "summary" => "Search Game Level by Billable Entity",
  "nickname" => "get_game_levels_by_billable_entity",
  "responseClass" => "GameLevelResponse",
  "endpoint" => "/api/{version}/level/searchByBillableEntity",
  "notes" => "Searches on game levels that the logged in user has access to. A user would have access if the creator of the game level is managed under the same BillableEntity.",
  "parameters" => [
    {
      "name" => "account_id",
      "description" => "The account id of the user",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "app_key",
      "description" => "the application key",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "keyword",
      "description" => "The keyword used to search",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "sort_field",
      "description" => "The column to sort the search on",
      "dataType" => "String",
      "allowableValues" => "[LEVEL_ACTIVE, LEVEL_NAME, LEVEL_DESCRIPTION, LEVEL_CREATED, LEVEL_UPDATED, LEVEL_LIKES, LEVEL_DISLIKES, LEVEL_NOTES, LEVEL_PLAYS, LEVEL_DOWNLOADS, LEVEL_QUITS, LEVEL_COMPLETES, LEVEL_VERSION, LEVEL_MISSION_TYPE, LEVEL_OWNER_DISPLAY, GAME_OWNER_DISPLAY, GAME_TITLE, GAME_DESCRIPTION, GAME_LIKES, GAME_DISLIKES, APP_NAME, APP_SCORING_TYPE]",
      "paramType" => "query",
    },
    {
      "name" => "descending",
      "description" => "The order to return the search results",
      "dataType" => "Boolean",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "active_only",
      "description" => "Return only active results",
      "dataType" => "Boolean",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "start",
      "description" => "The record to begin the return set on",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "limit",
      "description" => "The number of records to return",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "version",
      "description" => "",
      "dataType" => "Float",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('GET', '/api/{version}/level/questions/get', {
  "resourcePath" => "/GameLevel",
  "summary" => "Get Level Questions",
  "nickname" => "get_questions_in_level",
  "responseClass" => "QuestionResponse",
  "endpoint" => "/api/{version}/level/questions/get",
  "notes" => "Get questions within a level.",
  "parameters" => [
    {
      "name" => "level_id",
      "description" => "the id of the level to get questions from",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "account_id",
      "description" => "the id of the logged in user",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "version",
      "description" => "",
      "dataType" => "Float",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('GET', '/api/{version}/level/words/get', {
  "resourcePath" => "/GameLevel",
  "summary" => "Get Level Words",
  "nickname" => "get_words_in_level",
  "responseClass" => "WordzWordResponse",
  "endpoint" => "/api/{version}/level/words/get",
  "notes" => "Get words within a level.",
  "parameters" => [
    {
      "name" => "level_id",
      "description" => "the id of the level to get words for",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "account_id",
      "description" => "the id of the logged in user",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "version",
      "description" => "",
      "dataType" => "Float",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('POST', '/api/{version}/level/update', {
  "resourcePath" => "/GameLevel",
  "summary" => "Update Game Level",
  "nickname" => "update_game_level",
  "responseClass" => "GameLevelResponse",
  "endpoint" => "/api/{version}/level/update",
  "notes" => "Update a game level. Currently does NOT support game objects.",
  "parameters" => [
    {
      "name" => "account_id",
      "description" => "The logged in user.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "app_key",
      "description" => "The game application key to save the level for.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "level_id",
      "description" => "If update then include the level Id.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "name",
      "description" => "The name of the level.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "description",
      "description" => "The description of the level.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "difficulty",
      "description" => "The difficulty, possible values are: VERY_EASY, EASY, MEDIUM, HARD, VERY_HARD.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "app_version",
      "description" => "The version number of the applicatuion required to correctly load/play the level.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "asset_image_id",
      "description" => "The asset Id of the level image.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "asset_icon_id",
      "description" => "The asset Id of the level icon.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "game_data",
      "description" => "The game level data: xml, json, or other texted based format.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "game_data_suffix",
      "description" => "The game level data format type.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "visibility",
      "description" => "Is the level visible to others, possible values are: PUBLIC, PRIVATE.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "friend_group",
      "description" => "Make the level be readable by all friends.",
      "dataType" => "Boolean",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "connection_ids",
      "description" => "Make the level be readable by connections in this list.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "connection_group_ids",
      "description" => "Make the level be readable by connection groups in this list.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "balance",
      "description" => "Set the amount of money available to spend, once 0 deactivate level. Set to a negative value for unlimited.",
      "dataType" => "Float",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "active",
      "description" => "If true set the game level as active.",
      "dataType" => "Boolean",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "allocate_tickets",
      "description" => "If true then scoring will give tickets.",
      "dataType" => "Boolean",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "ticket_count",
      "description" => "The number of tickets to reward",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "ticket_type",
      "description" => "The type of ticket to reward, null means default type",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "points",
      "description" => "The number of points to award for completing a level",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "tutorial_title",
      "description" => "Title of the tutorial.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "tutorial_message",
      "description" => "Message of the tutorial.",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "tutorial_alignment",
      "description" => "Alignment of the tutorial image, possible values are: NONE, IMAGE_ABOVE, IMAGE_BELOW, IMAGE_LEFT, IMAGE_RIGHT, IMAGE_ONLY, TEXT_ONLY",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "tutorial_image_asset_id",
      "description" => "Asset id of the tutorial image.",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "offer_id",
      "description" => "",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "meta_data",
      "description" => "external custom client defined data",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "version",
      "description" => "",
      "dataType" => "Float",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('POST', '/api/{version}/level/questions/update', {
  "resourcePath" => "/GameLevel",
  "summary" => "Update Level Questions",
  "nickname" => "update_questions_in_level",
  "responseClass" => "SirqulResponse",
  "endpoint" => "/api/{version}/level/questions/update",
  "notes" => "Updates a level with question game objects.",
  "parameters" => [
    {
      "name" => "level_id",
      "description" => "the id of the level to update questions on",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "account_id",
      "description" => "the id of the logged in user",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "question_ids",
      "description" => "the IDs of the questions to update",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "version",
      "description" => "",
      "dataType" => "Float",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end


MyApp.add_route('POST', '/api/{version}/level/words/update', {
  "resourcePath" => "/GameLevel",
  "summary" => "Update Level Words",
  "nickname" => "update_words_in_level",
  "responseClass" => "SirqulResponse",
  "endpoint" => "/api/{version}/level/words/update",
  "notes" => "Updates a level with word game objects.",
  "parameters" => [
    {
      "name" => "level_id",
      "description" => "the id of the level to update words for",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "account_id",
      "description" => "the id of the logged in user",
      "dataType" => "Integer",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "word_ids",
      "description" => "the ids of the words to update for the level",
      "dataType" => "String",
      "allowableValues" => "",
      "paramType" => "query",
    },
    {
      "name" => "version",
      "description" => "",
      "dataType" => "Float",
      "paramType" => "path",
    },
    ]}) do
  cross_origin
  # the guts live here

  {"message" => "yes, it worked"}.to_json
end

File: server-scala-play-server/app/api/GameLevelApi.scala
package api

import play.api.libs.json._
import model.BigDecimal
import model.GameLevelListResponse
import model.GameLevelResponse
import model.QuestionResponse
import model.SirqulResponse
import model.WordzWordResponse

@javax.annotation.Generated(value = Array("org.openapitools.codegen.languages.ScalaPlayFrameworkServerCodegen"), date = "2026-02-09T20:57:39.376804970Z[Etc/UTC]", comments = "Generator version: 7.20.0-SNAPSHOT")
trait GameLevelApi {
  /**
    * Create Game Level
    * Create a game level. Currently does NOT support game objects.
    * @param accountId The logged in user.
    * @param name The name of the level.
    * @param gameData The game level data: xml, json, or other text based format.
    * @param gameDataSuffix The game level data format type.
    * @param appKey The game application key to save the level for.
    * @param description The description of the level.
    * @param difficulty The difficulty, possible values are: VERY_EASY, EASY, MEDIUM, HARD, VERY_HARD.
    * @param appVersion The version number of the application required to correctly load/play the level.
    * @param assetImageId The asset Id of the level image.
    * @param assetIconId The asset Id of the level icon.
    * @param visibility Is the level visible to others, possible values are: PUBLIC, PRIVATE.
    * @param friendGroup Make the level be readable by all friends.
    * @param connectionIds Make the level be readable by connections in this list.
    * @param connectionGroupIds Make the level be readable by connection groups in this list.
    * @param balance Set the amount of money available to spend, once 0 deactivate level. Set to a negative value for unlimited.
    * @param active If true set the game level as active. Default is false.
    * @param allocateTickets If true then scoring will give tickets. Default is false.
    * @param ticketCount The number of tickets to reward
    * @param ticketType The type of ticket to reward, null means default type
    * @param points The number of points to award for completing a level
    * @param tutorialTitle Title of the tutorial.
    * @param tutorialMessage Message of the tutotrial.
    * @param tutorialAlignment Alignment of the tutorial image. Default to NONE. Possible values are: NONE, IMAGE_ABOVE, IMAGE_BELOW, IMAGE_LEFT, IMAGE_RIGHT, IMAGE_ONLY, TEXT_ONLY
    * @param tutorialImageAssetId Asset id of the tutorial image.
    * @param offerId id of the offer
    * @param metaData external custom client defined data
    */
  def createGameLevel(version: BigDecimal, accountId: Long, name: String, gameData: String, gameDataSuffix: String, appKey: Option[String], description: Option[String], difficulty: Option[String], appVersion: Option[String], assetImageId: Option[Long], assetIconId: Option[Long], visibility: Option[String], friendGroup: Option[Boolean], connectionIds: Option[String], connectionGroupIds: Option[String], balance: Option[Double], active: Option[Boolean], allocateTickets: Option[Boolean], ticketCount: Option[Long], ticketType: Option[String], points: Option[Long], tutorialTitle: Option[String], tutorialMessage: Option[String], tutorialAlignment: Option[String], tutorialImageAssetId: Option[Long], offerId: Option[Long], metaData: Option[String]): GameLevelResponse

  /**
    * Delete Game Level
    * Delete a game level. The level and account must be valid and have the appropirate permissions to view the content.
    * @param accountId The logged in user.
    * @param levelId The id of the level to return.
    */
  def deleteGameLevel(version: BigDecimal, accountId: Long, levelId: Long): SirqulResponse

  /**
    * Get Game Level
    * Get a game level. The level and account must be valid and have the appropirate permissions to view the content.
    * @param accountId The logged in user.
    * @param levelId The id of the level to return.
    * @param includeGameData If true include the game level data, otherwise don&#39;t. default is false.
    */
  def getGameLevel(version: BigDecimal, accountId: Long, levelId: Long, includeGameData: Option[Boolean]): GameLevelResponse

  /**
    * Search Game Levels
    * Get a list of levels for an application, just those the account has permissions to view.
    * @param accountId The logged in user.
    * @param appKey the application key
    * @param keyword Match the keyword to the owner name or level name.
    * @param sortField what field to sort on
    * @param descending whether to return levels in ascending or descending order
    * @param start Start the result set at some index.
    * @param limit Limit the result to some number.
    * @param appVersion The maximum version of the level to return.
    * @param includeGameData If true include the game level data, otherwise don&#39;t. default is false.
    * @param filters 
    */
  def getGameLevelsByApplication(version: BigDecimal, accountId: Long, appKey: String, keyword: Option[String], sortField: Option[String], descending: Option[Boolean], start: Option[Int], limit: Option[Int], appVersion: Option[String], includeGameData: Option[Boolean], filters: Option[String]): GameLevelListResponse

  /**
    * Search Game Level by Billable Entity
    * Searches on game levels that the logged in user has access to. A user would have access if the creator of the game level is managed under the same BillableEntity.
    * @param accountId The account id of the user
    * @param appKey the application key
    * @param keyword The keyword used to search
    * @param sortField The column to sort the search on
    * @param descending The order to return the search results
    * @param activeOnly Return only active results
    * @param start The record to begin the return set on
    * @param limit The number of records to return
    */
  def getGameLevelsByBillableEntity(version: BigDecimal, accountId: Long, appKey: Option[String], keyword: Option[String], sortField: Option[String], descending: Option[Boolean], activeOnly: Option[Boolean], start: Option[Long], limit: Option[Long]): GameLevelResponse

  /**
    * Get Level Questions
    * Get questions within a level.
    * @param levelId the id of the level to get questions from
    * @param accountId the id of the logged in user
    */
  def getQuestionsInLevel(version: BigDecimal, levelId: Long, accountId: Long): QuestionResponse

  /**
    * Get Level Words
    * Get words within a level.
    * @param levelId the id of the level to get words for
    * @param accountId the id of the logged in user
    */
  def getWordsInLevel(version: BigDecimal, levelId: Long, accountId: Long): WordzWordResponse

  /**
    * Update Game Level
    * Update a game level. Currently does NOT support game objects.
    * @param accountId The logged in user.
    * @param levelId If update then include the level Id.
    * @param appKey The game application key to save the level for.
    * @param name The name of the level.
    * @param description The description of the level.
    * @param difficulty The difficulty, possible values are: VERY_EASY, EASY, MEDIUM, HARD, VERY_HARD.
    * @param appVersion The version number of the applicatuion required to correctly load/play the level.
    * @param assetImageId The asset Id of the level image.
    * @param assetIconId The asset Id of the level icon.
    * @param gameData The game level data: xml, json, or other texted based format.
    * @param gameDataSuffix The game level data format type.
    * @param visibility Is the level visible to others, possible values are: PUBLIC, PRIVATE.
    * @param friendGroup Make the level be readable by all friends.
    * @param connectionIds Make the level be readable by connections in this list.
    * @param connectionGroupIds Make the level be readable by connection groups in this list.
    * @param balance Set the amount of money available to spend, once 0 deactivate level. Set to a negative value for unlimited.
    * @param active If true set the game level as active.
    * @param allocateTickets If true then scoring will give tickets.
    * @param ticketCount The number of tickets to reward
    * @param ticketType The type of ticket to reward, null means default type
    * @param points The number of points to award for completing a level
    * @param tutorialTitle Title of the tutorial.
    * @param tutorialMessage Message of the tutorial.
    * @param tutorialAlignment Alignment of the tutorial image, possible values are: NONE, IMAGE_ABOVE, IMAGE_BELOW, IMAGE_LEFT, IMAGE_RIGHT, IMAGE_ONLY, TEXT_ONLY
    * @param tutorialImageAssetId Asset id of the tutorial image.
    * @param offerId 
    * @param metaData external custom client defined data
    */
  def updateGameLevel(version: BigDecimal, accountId: Long, levelId: Long, appKey: Option[String], name: Option[String], description: Option[String], difficulty: Option[String], appVersion: Option[String], assetImageId: Option[Long], assetIconId: Option[Long], gameData: Option[String], gameDataSuffix: Option[String], visibility: Option[String], friendGroup: Option[Boolean], connectionIds: Option[String], connectionGroupIds: Option[String], balance: Option[Double], active: Option[Boolean], allocateTickets: Option[Boolean], ticketCount: Option[Long], ticketType: Option[String], points: Option[Long], tutorialTitle: Option[String], tutorialMessage: Option[String], tutorialAlignment: Option[String], tutorialImageAssetId: Option[Long], offerId: Option[Long], metaData: Option[String]): GameLevelResponse

  /**
    * Update Level Questions
    * Updates a level with question game objects.
    * @param levelId the id of the level to update questions on
    * @param accountId the id of the logged in user
    * @param questionIds the IDs of the questions to update
    */
  def updateQuestionsInLevel(version: BigDecimal, levelId: Long, accountId: Long, questionIds: String): SirqulResponse

  /**
    * Update Level Words
    * Updates a level with word game objects.
    * @param levelId the id of the level to update words for
    * @param accountId the id of the logged in user
    * @param wordIds the ids of the words to update for the level
    */
  def updateWordsInLevel(version: BigDecimal, levelId: Long, accountId: Long, wordIds: String): SirqulResponse
}
